### PR TITLE
Simplify statsd pattern parser

### LINKED ISSUE #1100: statsd monitor converter fails to parse `pattern` or `metricName` values not ending with a {pattern}
signalfx-agent config file -> monitors -> statsd -> converters:

If `pattern` or `metricName` doesn't have any {patterns}, for example `metric.count`, signalfx agent ignores the rule with the following error message in logs:

```
level=error msg="Invalid pattern. Mismatched brackets : metric.count" monitorType=statsd
```

If `pattern` or `metricName` does have {patterns} but not at the end for of the string, for example `metric.{region}.count`, signalfx-agent fails to start with the following stacktrace:

```
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: panic: runtime error: slice bounds out of range [39:38]
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: goroutine 97 [running]:
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: github.com/signalfx/signalfx-agent/internal/monitors/statsd.parseFields(0xc000ce47e0, 0x28, 0xc0009738e0)
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/src/signalfx-agent/internal/monitors/statsd/parser.go:54 +0x5db
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: github.com/signalfx/signalfx-agent/internal/monitors/statsd.initConverter(0xc000a3d128, 0xc00003e018)
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/src/signalfx-agent/internal/monitors/statsd/parser.go:18 +0x3b
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: github.com/signalfx/signalfx-agent/internal/monitors/statsd.(*Monitor).Configure(0xc000cfd4a0, 0xc00035f400, 0x0, 0x0)
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/src/signalfx-agent/internal/monitors/statsd/monitor.go:90 +0x168
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: reflect.Value.call(0x1b748c0, 0xc000cfd4a0, 0x213, 0x1e23a00, 0x4, 0xc000a3d520, 0x1, 0x1, 0x5, 0xc000a3d418, ...)
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/local/go/src/reflect/value.go:460 +0x5f6
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: reflect.Value.Call(0x1b748c0, 0xc000cfd4a0, 0x213, 0xc000a3d520, 0x1, 0x1, 0xc000cfd4a0, 0x213, 0xc000cfd4a0)
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/local/go/src/reflect/value.go:321 +0xb4
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: github.com/signalfx/signalfx-agent/internal/core/config.CallConfigure(0x1b748c0, 0xc000cfd4a0, 0x1cc4760, 0xc00035f400, 0x30, 0xc000cfd8c0)
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/src/signalfx-agent/internal/core/config/dynamic.go:111 +0x369
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: github.com/signalfx/signalfx-agent/internal/monitors.(*ActiveMonitor).configureMonitor(0xc00065a690, 0x21ba7a0, 0xc00035f400, 0x0, 0x21ba7a0)
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/src/signalfx-agent/internal/monitors/activemonitor.go:87 +0x381
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: github.com/signalfx/signalfx-agent/internal/monitors.(*MonitorManager).createAndConfigureNewMonitor(0xc000222f80, 0x21ba7a0, 0xc00035f040, 0x0, 0x0, 0xc000a3d868, 0x7b216d9a5a6a2bf1)
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/src/signalfx-agent/internal/monitors/manager.go:386 +0x6e5
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: github.com/signalfx/signalfx-agent/internal/monitors.(*MonitorManager).handleNewConfig(0xc000222f80, 0xc000cee400, 0xc1ca85db204e7540, 0xc00044fc48, 0x0, 0x0)
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/src/signalfx-agent/internal/monitors/manager.go:179 +0xa5
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: github.com/signalfx/signalfx-agent/internal/monitors.(*MonitorManager).Configure(0xc000222f80, 0xc000213400, 0xc, 0xc, 0xc0005bf208, 0xa, 0xc0009ac201)
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/src/signalfx-agent/internal/monitors/manager.go:106 +0x3a2
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: github.com/signalfx/signalfx-agent/internal/core.(*Agent).configure(0xc000222f00, 0xc0005bf000)
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/src/signalfx-agent/internal/core/agent.go:134 +0x3e9
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: github.com/signalfx/signalfx-agent/internal/core.Startup.func1(0xc0009610e0, 0x1e4ac0d, 0x18, 0xc000222f00, 0xc00092d200, 0x2206140, 0xc000130b40)
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/src/signalfx-agent/internal/core/agent.go:191 +0x172
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]: created by github.com/signalfx/signalfx-agent/internal/core.Startup
Dec 03 01:30:18 ip-xx-xx-xxx-xx signalfx-agent[24212]:         /usr/src/signalfx-agent/internal/core/agent.go:178 +0x147
```

### PR DESCRIPTION
**Pull request by keitwb**
This makes it a simple for loop/state machine pattern

Fixes #1100

### CODE REVIEW

File: internal/monitors/statsd/parser_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+			true,
+			false,
+		},
+		{
+			// Cannot have back-to-back patterns
+			"{cluster}.cds_{traffic}{mesh}_{service}-vn_{}.{action}",
+			nil,
+			false,
+			true,
+		},
```
Comment (dmitryax) [inline internal/monitors/statsd/parser_test.go]: Can we have a test case with for a string not ending with a {pattern}? like `metric.{region}.count`?

File: docs/monitors/statsd.md [hunk 2]
```diff
@@ -15,16 +15,15 @@ datapoints.  It listens on a configured address and port in order to
...
 `gauge` respectively.
 
 **Note that datapoints will get a `host` dimension of the current host that
-the agent is running on, not the host from which the statsd metric was
-sent.  For this reason, it is recommended to send statsd metrics to a local
-agent instance.**
+the agent is running on, not the host from which the statsd metric was sent.
+For this reason, it is recommended to send statsd metrics to a local agent
+instance. If you don't want the `host` dimension, you can set
+`disableHostDimensions: true` on the monitor configuration**
```
Comment (dmitryax) [inline docs/monitors/statsd.md]: Could you please update the example in the _Formatting metric name_ section: `metric: "{traffic}.{action}"` -> `metricName: "{traffic}.{action}"`

**Comment by dmitryax** [comment]
Thank you for fixing it so quickly!

File: docs/monitors/statsd.md (hunk 2 above)
Comment (keitwb) [reply docs/monitors/statsd.md]: Fixed.

File: internal/monitors/statsd/parser_test.go (hunk 1 above)
Comment (keitwb) [reply internal/monitors/statsd/parser_test.go]: Added two cases with only inner dynamic fields, one with a pattern that starts with a dynamic field, and another that doesn't.

Comment (dmitryax) [reply internal/monitors/statsd/parser_test.go]: Cool, thanks!

**Comment by keitwb** [comment]
Thanks for submitting the issue @dmitryax.  I will release this as a patch release later today or tomorrow, whenever I get a chance.

**Pull request merged**